File: shared/physique.js
"use strict";

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

function estimateMeasurements(sex, heightCm, weightKg) {
  const bmi = weightKg / Math.pow(heightCm / 100, 2);
  const neck = sex === "male" ? heightCm * 0.22 : heightCm * 0.19;
  const waist = sex === "male"
    ? heightCm * (0.43 + Math.max(0, bmi - 22) * 0.008)
    : heightCm * (0.39 + Math.max(0, bmi - 21) * 0.008);
  const hips = sex === "female"
    ? heightCm * (0.55 + Math.max(0, bmi - 21) * 0.006)
    : heightCm * 0.52;

  return {
    neck: Math.round(neck),
    waist: Math.round(waist),
    hips: Math.round(hips)
  };
}

function calculatePhysique(input) {
  const sex = input.sex || input.gender || "male";
  const height = Number(input.heightCm || input.height);
  const weight = Number(input.weightKg || input.weight);
  const goal = Number(input.goalBodyFat || input.goal || 14);
  const estimated = estimateMeasurements(sex, height, weight);
  const neck = Number(input.neckCm || input.neck || estimated.neck);
  const waist = Number(input.waistCm || input.waist || estimated.waist);
  const hips = Number(input.hipsCm || input.hips || estimated.hips);

  if (!Number.isFinite(height) || height < 90 || height > 230) {
    throw new Error("heightCm must be between 90 and 230");
  }
  if (!Number.isFinite(weight) || weight < 30 || weight > 260) {
    throw new Error("weightKg must be between 30 and 260");
  }

  let bodyFat;
  if (sex === "male") {
    bodyFat = 495 / (1.0324 - 0.19077 * Math.log10(waist - neck) + 0.15456 * Math.log10(height)) - 450;
  } else {
    bodyFat = 495 / (1.29579 - 0.35004 * Math.log10(waist + hips - neck) + 0.22100 * Math.log10(height)) - 450;
  }

  bodyFat = clamp(bodyFat, 3, 60);
  const bmi = weight / Math.pow(height / 100, 2);
  const fatMassKg = weight * bodyFat / 100;
  const leanMassKg = weight - fatMassKg;
  const targetWeightKg = leanMassKg / (1 - goal / 100);
  const toGoalKg = targetWeightKg - weight;

  return {
    bodyFat,
    bmi,
    fatMassKg,
    leanMassKg,
    targetWeightKg,
    toGoalKg,
    goalBodyFat: goal,
    measurements: { neckCm: neck, waistCm: waist, hipsCm: hips }
  };
}

module.exports = {
  calculatePhysique,
  estimateMeasurements
};

File: apps/api/app/services/physique.py
from math import log10


def clamp(value: float, minimum: float, maximum: float) -> float:
    return min(maximum, max(minimum, value))


def estimate_measurements(sex: str, height_cm: float, weight_kg: float) -> dict:
    bmi = weight_kg / ((height_cm / 100) ** 2)
    neck = height_cm * (0.22 if sex == "male" else 0.19)
    if sex == "male":
        waist = height_cm * (0.43 + max(0, bmi - 22) * 0.008)
        hips = height_cm * 0.52
    else:
        waist = height_cm * (0.39 + max(0, bmi - 21) * 0.008)
        hips = height_cm * (0.55 + max(0, bmi - 21) * 0.006)

    return {
        "neck_cm": round(neck),
        "waist_cm": round(waist),
        "hips_cm": round(hips)
    }


def calculate_physique(payload: dict) -> dict:
    sex = payload.get("sex", "male")
    height = float(payload["height_cm"])
    weight = float(payload["weight_kg"])
    goal = float(payload.get("goal_body_fat", 14))
    estimated = estimate_measurements(sex, height, weight)
    neck = float(payload.get("neck_cm") or estimated["neck_cm"])
    waist = float(payload.get("waist_cm") or estimated["waist_cm"])
    hips = float(payload.get("hips_cm") or estimated["hips_cm"])

    if not 90 <= height <= 230:
        raise ValueError("height_cm must be between 90 and 230")
    if not 30 <= weight <= 260:
        raise ValueError("weight_kg must be between 30 and 260")

    if sex == "male":
        body_fat = 495 / (1.0324 - 0.19077 * log10(waist - neck) + 0.15456 * log10(height)) - 450
    else:
        body_fat = 495 / (1.29579 - 0.35004 * log10(waist + hips - neck) + 0.22100 * log10(height)) - 450

    body_fat = clamp(body_fat, 3, 60)
    bmi = weight / ((height / 100) ** 2)
    fat_mass_kg = weight * body_fat / 100
    lean_mass_kg = weight - fat_mass_kg
    target_weight_kg = lean_mass_kg / (1 - goal / 100)

    return {
        "body_fat": body_fat,
        "bmi": bmi,
        "fat_mass_kg": fat_mass_kg,
        "lean_mass_kg": lean_mass_kg,
        "target_weight_kg": target_weight_kg,
        "to_goal_kg": target_weight_kg - weight,
        "goal_body_fat": goal,
        "measurements": {
            "neck_cm": neck,
            "waist_cm": waist,
            "hips_cm": hips
        }
    }
